salt-cloud: Do not pass userdata_file through yaml renderer

terminalmage opened the pull request:
2016.11.0 added support for templating the ``userdata_file`` in the ec2 salt-cloud driver (see #32698). However, this feature used the master's ``renderer`` config option, which first evaluates jinja templating, followed by passing the data through the YAML loader. The problem with doing this, however, is that PyYAML does unpredictable things when you attempt to load non-YAML data through it. For example:

```python
% python
Python 2.7.13 (default, Dec 21 2016, 07:16:46)
[GCC 6.2.1 20160830] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> x = '''\
... #!/bin/bash
...
... echo "a" >> /var/tmp/a
... echo "b" >> /var/tmp/b
... echo "c" >> /var/tmp/c
... echo "d" >> /var/tmp/d
...
... echo "e" >> /var/tmp/e
... echo "f" >> /var/tmp/f
...
... echo "g" >> /var/tmp/g
... echo "h" >> /var/tmp/h
... '''
>>> import yaml
>>> print(x)

echo "a" >> /var/tmp/a
echo "b" >> /var/tmp/b
echo "c" >> /var/tmp/c
echo "d" >> /var/tmp/d

echo "e" >> /var/tmp/e
echo "f" >> /var/tmp/f

echo "g" >> /var/tmp/g
echo "h" >> /var/tmp/h

>>> print(yaml.safe_load(x))
echo "a" >> /var/tmp/a echo "b" >> /var/tmp/b echo "c" >> /var/tmp/c echo "d" >> /var/tmp/d
echo "e" >> /var/tmp/e echo "f" >> /var/tmp/f
echo "g" >> /var/tmp/g echo "h" >> /var/tmp/h
>>>
```

This PR fixes the issue by only templating when explicitly configured to do so. It does this by adding an optional cloud profile config paramter called ``userdata_template``, designed to work like the ``template`` param in ``file.managed`` states. For example:

```yaml
my-ec2-config:
  # Pass userdata to the instance to be created
  userdata_file: /etc/salt/my-userdata-file
  userdata_template: jinja
```

This option can also be set globally in the master configuration file.

Additionally, the other cloud drivers which support ``userdata_file`` (azurearm, nova, openstack) have had templating support added.

Finally, this PR addresses the fact that jinja will eat Windows newlines and spit out unix newlines:

```python
>>> import jinja2
>>> jinja2.Template('foo\r\nbar\r\nbaz\r\n').render()
u'foo\nbar\nbaz'
```

It fixes this by checking the template input for Windows newlines (``\r\n``) before processing the render pipe, and then checking the rendered template afterwards and replacing the newlines with Windows newlines if necessary.


CC: @techhat

terminalmage committed:
Preserve windows newlines in salt.template.compile_template()

Test included.

terminalmage committed:
Add userdata_renderer master config param

terminalmage committed:
Add documentation for userdata_renderer

terminalmage committed:
Add userdata_renderer fix info to 2016.11.4 release notes

terminalmage committed:
Openstack did not have templating support for userdata_file before 2016.11.4

terminalmage committed:
ec2: Add support for using userdata_renderer to template userdata_file

terminalmage committed:
Add templating support for other cloud drivers that support userdata_file

terminalmage committed:
Remove userdata_renderer value

terminalmage committed:
Rename userdata_renderer -> userdata_template in master config docs

terminalmage committed:
Update cloud docs to reflect userdata_renderer -> userdata_template

terminalmage committed:
Only template the userdata_file if explicitly configured to do so

terminalmage committed:
Try to read compiled template as StringIO

The fact that PyYAML was the last renderer in the render pipe initially,
masked the fact that template renderers return StringIO instances and
not strings. This commit makes sure that we try reading from the
StringIO first and then fallback to just trying to base64 encode the
userdata value as returned by compile_template().

terminalmage committed:
Update 2016.11.4 release notes for userdata_renderer -> userdata_template

terminalmage committed:
Move userdata templating to salt.utils.cloud

terminalmage committed:
Bring in salt.utils.stringio from develop branch

terminalmage committed:
Properly handle renderers which return StringIO objects

terminalmage committed:
Update compile_template test to use StringIO

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @techhat, @trevor-h and @pabelanger to be potential reviewers.

terminalmage committed:
Allow for userdata_template to be disabled in a cloud_profile

This handles the case where a global value has been set and it is
necessary to disable it for a single profile.

terminalmage commented:
@gtmanfred helped me find an obvious syntax error, which I've just pushed a fix for.

terminalmage committed:
Nova and openstack don't accept base64-encoded userdata

This fixes my bad copypasta.

terminalmage commented:
Rebased to fix merge conflict in release notes.
